Self-Consistent-Field/RHF_EFP.py

```diff
@@ -0,0 +1,376 @@
+
+__copyright__ = "(c) 2014-2017, The Psi4NumPy Developers"
+__license__   = "BSD-3-Clause"
+__date__      = "2017-08-28"
+
+import time
+import numpy as np
+np.set_printoptions(precision=5, linewidth=200, suppress=True)
+import psi4
+import pylibefp
+
+import os
+
+
+# Memory for Psi4 in GB
+psi4.set_memory('500 MB')
+psi4.core.set_output_file("output.dat", False)
+
+# Memory for numpy in GB
+numpy_memory = 2
+
+def set_qm_atoms(mol, efpobj):
+    """Provides list of coordinates of quantum mechanical atoms from
+    psi4.core.Molecule `mol` to pylibefp.core.efp() `efpobj`.
+
+    """
+    ptc = []
+    coords = []
+    for iat in range(mol.natom()):
+        ptc.append(mol.charge(iat))
+        coords.append(mol.x(iat))
+        coords.append(mol.y(iat))
+        coords.append(mol.z(iat))
+
+    efpobj.set_point_charges(ptc, coords)
+
+
+def modify_Fock_permanent(mol, nbf, efpobj):
+    """Computes array of the EFP contribution to the potential felt by
+    QM atoms, due to permanent EFP moments, for a SCF procedure.
+
+    Requires psi4.core.Molecule `mol`, number of basis functions `nbf`,
+    and pylibefp.core.efp() `efpobj`.
+
+    """
+    # get composition counts from libefp
+    n_fr = efpobj.get_frag_count()
+    natoms = efpobj.get_frag_atom_count()
+
+    # get multipoles count, pos'n, values from libefp
+    #   charge + dipoles + quadrupoles + octupoles = 20
+    n_mp = efpobj.get_multipole_count()
+    xyz_mp = np.asarray(efpobj.get_multipole_coordinates()).reshape(n_mp, 3)
+    val_mp = np.asarray(efpobj.get_multipole_values()).reshape(n_mp, 20)
+
+    #                    0  X  Y  Z  XX   YY   ZZ   XY   XZ   YZ
+    prefacs = np.array([ 1, 1, 1, 1, 1/3, 1/3, 1/3, 2/3, 2/3, 2/3,
+        1/15, 1/15, 1/15, 3/15, 3/15, 3/15, 3/15, 3/15, 3/15, 6/15])
+    #   XXX   YYY   ZZZ   XXY   XXZ   XYY   YYZ   XZZ   YZZ   XYZ
+
+    # EFP permanent moment contribution to the Fock Matrix
+    V2 = np.zeros((nbf, nbf))
+
+    # Cartesian basis one-electron EFP perturbation
+    efp_ints = np.zeros((20, nbf, nbf))
+
+    for imp in range(n_mp):
+        origin = xyz_mp[imp]
+
+        # get EFP multipole integrals from Psi4
+        p4_efp_ints = mints.ao_efp_multipole_potential(origin=origin)
+        for pole in range(20):
+            efp_ints[pole] = np.asarray(p4_efp_ints[pole])
+
+        # add frag atom Z into multipole charge (when pos'n of atom matches mp)
+        for ifr in range(n_fr):
+            atoms = efpobj.get_frag_atoms(ifr)
+            for iat in range(natoms[ifr]):
+                xyz_atom = [atoms[iat]['x'], atoms[iat]['y'], atoms[iat]['z']]
+                if np.allclose(xyz_atom, origin, atol=1e-10):
+                    val_mp[imp, 0] += atoms[iat]['Z']
+
+        # scale multipole integrals by multipole magnitudes. result goes into V
+        for pole in range(20):
+            efp_ints[pole] *= -prefacs[pole] * val_mp[imp, pole]
+            V2 += efp_ints[pole]
```

Comment:
Can always use broadcasting here. If I understand it correctly:

`V2 = np.sum(- efp_ints * val_mp[:, None], axis=0)`

Reply:
I don't think so, as V2 is accumulated across both imp and pole loops, while efp_ints is redefined for each imp?